src/EditorFeatures/Core/ValueTracking/ValueTrackingService.cs

```diff
@@ -23,101 +35,355 @@ public ValueTrackingService()
+                                    symbol,
+                                    document.Project.Solution,
+                                    findReferenceProgressCollector,
+                                    documents: null, FindReferencesSearchOptions.Default, cancellationToken).ConfigureAwait(false);
         }
 
-        private class FindReferencesProgress : IStreamingFindReferencesProgress, IStreamingProgressTracker
+        private static async Task TrackParameterSymbolAsync(
+            IParameterSymbol parameterSymbol,
+            Document document,
+            ValueTrackingProgressCollector progressCollector,
+            CancellationToken cancellationToken)
         {
-            private readonly ValueTrackingProgressCollector _valueTrackingProgressCollector;
-            public FindReferencesProgress(ValueTrackingProgressCollector valueTrackingProgressCollector)
+            var containingMethod = (IMethodSymbol)parameterSymbol.ContainingSymbol;
+            var findReferenceProgressCollector = new FindReferencesProgress(progressCollector);
+            await SymbolFinder.FindReferencesAsync(
+                containingMethod,
+                document.Project.Solution,
+                findReferenceProgressCollector,
+                documents: null, FindReferencesSearchOptions.Default, cancellationToken).ConfigureAwait(false);
+        }
+
+        private static async Task TrackMethodSymbolAsync(IMethodSymbol methodSymbol, Document document, ValueTrackingProgressCollector progressCollector, CancellationToken cancellationToken)
+        {
+            var hasAnyOutData = HasAValueReturn(methodSymbol) || HasAnOutOrRefParam(methodSymbol);
+            if (!hasAnyOutData)
             {
-                _valueTrackingProgressCollector = valueTrackingProgressCollector;
+                // With no out data, there's nothing to do here
+                return;
             }
 
-            public IStreamingProgressTracker ProgressTracker => this;
+            // TODO: Use DFA to find meaningful returns? https://github.com/dotnet/roslyn-analyzers/blob/9e5f533cbafcc5579e4d758bc9bde27b7611ca54/docs/Writing%20dataflow%20analysis%20based%20analyzers.md 
+            if (HasAValueReturn(methodSymbol))
+            {
+                foreach (var location in methodSymbol.GetDefinitionLocationsToShow())
+                {
+                    if (location.SourceTree is null)
+                    {
+                        continue;
+                    }
 
-            public ValueTask AddItemsAsync(int count) => new();
+                    var node = location.FindNode(cancellationToken);
+                    var sourceDoc = document.Project.Solution.GetRequiredDocument(location.SourceTree);
+                    var syntaxFacts = sourceDoc.GetRequiredLanguageService<ISyntaxFactsService>();
+                    var returnStatements = node.DescendantNodesAndSelf().Where(n => syntaxFacts.IsReturnStatement(n)).ToImmutableArray();
```

Comment:
Does this need to ensure it doesn't descend into nested local functions or lambdas?

Reply:
Not sure yet, after finishing up some of the expression and method call situations lambdas and locals are next